desc/utils.py

```diff
@@ -431,6 +431,41 @@ def svd_inv_null(A):
     return Ainv, Z
 
 
+def qr_xp_null(A, b, tol=1e-10):
+    """Compute null space of a matrix and particular solution Ax=b using QR.
+
+    Parameters
+    ----------
+    A : ndarray
+        Matrix to find null space of.
+    b : ndarray
+        Right-hand side of Ax = b.
+
+    Returns
+    -------
+    x_p : ndarray
+        Particular solution to Ax = b.
+    Z : ndarray
+        Null space of A such that AZ=0 and Z.T@Z=I.
+
+    """
+    # Linear constraint matrix A is usually wide
+    # QR decomposition of A^T
+    Q, R = qr(A.T)
+    # Determine rank
+    diag = jnp.abs(jnp.diag(R))
+    rank = jnp.sum(diag > tol)
```

Comment:
This is a naive way to find the rank, but it seems to be working. I don't know if our linear constraints result in this kind of matrix but, we may need to consider,

$$
R = \begin{bmatrix}
x & x & x & x & x \\
0 & 0 & x & x & x \\
0 & 0 & 0 & 0 & x \\
0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 \\
\end{bmatrix}
$$

which should have rank=3 but my current implementation will give 1.

Reply:
Have you checked this against rank from SVD for some examples in DESC? this is an approximate way to get the rank... I am not comfortable doing this to grab the matrix rank here.

Reply:
you need to use pivoting qr to get the rank, the qr that you're using doesn't support that. try the scipy one

Reply:
should also add a test that this works correctly for linearly dependent constraints. The unique logic will remove actual duplicates but won't catch stuff like

x1 + x2 = 4
2*x1 + 2*x2 = 8

Reply:
I will try the scipy one when I profile again. But my hesitation is due to not being able to use GPU.

Reply:
Nevermind, I just use JAX svd ans it is fast enough.

Reply:
I can still try QR but without JAX implementing RRQR I don't think it can be performant. Because scipy thing is slow.